### PR TITLE
discord: add option to install Vencord

### PR DESCRIPTION
**Pull request by Scrumplex**
This PR moves `vencord-web-extension` from `webcord` to a top level package, allows building the injector scripts, and adds an option to `discord` to install Vencord.

I am unsure if `vencord` and `vencord-web-extension` should be top-level packages, as they are not really packages.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Scrumplex** [comment]
CC @Infinidoge @MP2E @Artturin for Discord

File: pkgs/development/misc/vencord/default.nix [hunk 1]
```diff
@@ -48,13 +49,13 @@ buildNpmPackage rec {
   ];
 
   installPhase = ''
-    cp -r dist/extension-unpacked $out
```
Comment (FlafyDev) [inline pkgs/development/misc/vencord/default.nix:51]: maybe change it to:
`cp -r dist/ $out` for build
`cp -r dist/extension-unpacked $out` for buildWeb

This way only the necessary files for vencord web extension will be installed and there will be no need for the `/extension-unpacked` part in
`vencord = "${vencord-web-extension}/extension-unpacked";` (`webcord-vencord/default.nix`)

**Review by FlafyDev — CHANGES_REQUESTED** [review]
Is `development/misc` the best directory for `vencord`?
I think `misc` would fit more, since I am not aware of any similar packages..

**Review by FlafyDev — APPROVED** [review]

**Comment by Infinidoge** [comment]
Has this been tested in combination with OpenASAR?

**Comment by Scrumplex** [comment]
> Has this been tested in combination with OpenASAR?

It works just as well. Vencord will just load the entrypoint in the _app.asar file, no matter if it's OpenASAR or the original.

File: pkgs/applications/networking/instant-messengers/discord/default.nix [hunk 2]
```diff
@@ -59,7 +59,7 @@ let
     downloadPage = "https://discordapp.com/download";
     sourceProvenance = with sourceTypes; [ binaryNativeCode ];
     license = licenses.unfree;
-    maintainers = with maintainers; [ MP2E artturin infinidoge jopejoe1 ];
+    maintainers = with maintainers; [ MP2E Scrumplex artturin infinidoge jopejoe1 ];
```
Comment (Infinidoge) [inline pkgs/applications/networking/instant-messengers/discord/default.nix:62]: Nit: This should technically be in its own commit

File: pkgs/misc/vencord/default.nix [hunk 3]
```diff
@@ -47,14 +48,16 @@ buildNpmPackage rec {
...
+  '' else ''
+    cp -r dist/ $out
   '';
 
   meta = with lib; {
     description = "Vencord web extension";
     homepage = "https://github.com/Vendicated/Vencord";
     license = licenses.gpl3Only;
-    maintainers = with maintainers; [ FlafyDev NotAShelf ];
+    maintainers = with maintainers; [ FlafyDev NotAShelf Scrumplex ];
```
Comment (Infinidoge) [inline pkgs/misc/vencord/default.nix:61]: Nit: This should be in its own commit

**Review by Infinidoge — CHANGES_REQUESTED** [review]
Overall, great PR! Everything seems to be working on my machine.
Just a couple nitpicks about commits:
- To my knowledge, adding yourself as a maintainer to a package should be done as their own commits.
- It might be a good idea to separate the refactoring of webcord-vencord into its own commit, though that I'm unsure about. It would definitely be a good idea to explain in the commit message why things were moved around, though.

**Review by Infinidoge — CHANGES_REQUESTED** [review]
Tiniest of nitpicks: Since Vencord is becoming the top level package, it should probably be `vencord: init ...` rather than `vencord-web-extension: init ...`, since later commits will be `vencord: ... -> ...`

**Comment by Scrumplex** [comment]
Yeah I wasn't sure about that part. Technically at that point I only add `vencord-web-extension` and later just refactor it to provide both `vencord` and `vencord-web-extension`

File: pkgs/applications/networking/instant-messengers/discord/linux.nix [hunk 4]
```diff
@@ -9,7 +9,7 @@
 , speechd
 , wayland
 , branch
-, common-updater-scripts, withOpenASAR ? false, withTTS ? false }:
+, common-updater-scripts, withOpenASAR ? false, withVencord ? false, withTTS ? false }:
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/discord/linux.nix:12]:
```suggestion
, common-updater-scripts
, withOpenASAR ? false
, withVencord ? false, vencord
, withTTS ? false
}:
```

File: pkgs/applications/networking/instant-messengers/discord/linux.nix [hunk 5]
```diff
@@ -1,4 +1,4 @@
-{ pname, version, src, openasar, meta, binaryName, desktopName, autoPatchelfHook
+{ pname, version, src, openasar, vencord, meta, binaryName, desktopName, autoPatchelfHook
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/instant-messengers/discord/linux.nix:1]:
```suggestion
{ pname, version, src, openasar, meta, binaryName, desktopName, autoPatchelfHook
```

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -24195,6 +24195,10 @@ with pkgs;
 
   vcg = callPackage ../development/libraries/vcg { };
 
+  vencord = callPackage ../misc/vencord { };
+
+  vencord-web-extension = callPackage ../misc/vencord { buildWebExtension = true; };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:24200]: How about we move this to passthru?

Comment (Scrumplex) [reply pkgs/top-level/all-packages.nix:24200]: Not quite sure. These are really two different build modes at the end.

`buildWebExtension = true;` will only build the web extension (for webcord), while `buildWebExtension = false;` will only build the variant for Discord

**Comment by Nyabinary** [comment]
@Scrumplex Vencord is outdated I think
